rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -357,24 +356,27 @@ class MatchArmTree extends ControlFlowTree instanceof MatchArm {
 
 class MatchExprTree extends PostOrderTree instanceof MatchExpr {
   override predicate propagatesAbnormal(AstNode child) {
-    child = [super.getExpr(), super.getMatchArmList().getAnArm().getExpr()]
+    child = [super.getExpr(), super.getAnArm().getExpr()]
   }
 
   override predicate first(AstNode node) { first(super.getExpr(), node) }
 
   override predicate succ(AstNode pred, AstNode succ, Completion c) {
     // Edge from the scrutinee to the first arm.
-    last(super.getExpr(), pred, c) and succ = super.getMatchArmList().getArm(0).getPat()
+    last(super.getExpr(), pred, c) and succ = super.getArm(0).getPat()
     or
     // Edge from a failed match/guard in one arm to the beginning of the next arm.
     exists(int i |
-      last(super.getMatchArmList().getArm(i), pred, c) and
-      first(super.getMatchArmList().getArm(i + 1), succ) and
+      last(super.getArm(i), pred, c) and
+      first(super.getArm(i + 1), succ) and
       c.(ConditionalCompletion).failed()
     )
     or
     // Edge from the end of each arm to the match expression.
-    last(super.getMatchArmList().getArm(_), pred, c) and succ = this and completionIsNormal(c)
+    last(super.getArm(_).getExpr(), pred, c) and succ = this and completionIsNormal(c)
```

Comment:
Isn't this better? Only simple outgoing edges from a match-arm should connect to the match expression, right?
```suggestion
    last(super.getArm(_), pred, c) and succ = this and c instanceof SimpleCompletion
```

Reply:
Eventually this can also be `BooleanCompletions`. For example, in
```rust
    fn test_nested_if_match(a: i64, b: i64) -> i64 {
        if (match a { 0 => b > 0, _ => b < 0 }) {
            1
        } else {
            0
        }
    }
```
both expressions `b > 0` and `b < 0` will have true/false edges.

Reply:
True, I guess to make it work the way I propose the `MatchArmTree` should be post-order. That way the true/false edges flow to the `MatchArm` node and the `MatchArm` -> `MatchExpr` step is a simple edge.

In general I prefer the flow of each node to be defined in terms of only its direct children without looking "deeper" (like peeking at the `getExpr()` of an "arm"). That way things are more composable. The way I see things a `MatchArm` node should have two normal outgoing edges a `no-match` and a `simple` one. The `no-match` ones are connected to the next arm and the `simple` ones connect to the `MatchExpr`.

On the other hand, it may be useful to deviate from that rule to cut away some edges and make the graph more compact.